### DISCUSSION TITLE
Is there a way to use React Query Hydration Boundary?

### DISCUSSION THREAD

**Discussion by DavideCarvalho** [trpc/trpc#5381] [Q&A]
Wanted to know if there is a way to use @tanstack/query [Hydration Boundary](https://tanstack.com/query/latest/docs/react/guides/ssr) to fetch data on the server side.

This approach is awesome since we don't need to change any code on the client side and just wrap existing components in a server component that uses it

**Comment by DavideCarvalho** [comment]
Sorry for mentioning you @KATT, but do you know if this is possible?

**Reply by niklaswallerstedt** [reply]
I've gotten the hydration boundary working on v11@next

**Reply by DavideCarvalho** [reply]
Hey @niklaswallerstedt ty for answering! Do you have a public repo or a codesandbox so I could take a look how you did it!

**Reply by niklaswallerstedt** [reply]
@DavideCarvalho sorry nothing public, but I can try and share some snippets later

**Comment by dihmeetree** [comment]
Here's what i've come up with: @DavideCarvalho
```ts
import { Suspense } from "react";
import { dehydrate, HydrationBoundary } from "@tanstack/react-query";
import { createServerSideHelpers } from "@trpc/react-query/server";
import superjson from "superjson";

import { appRouter } from "@acme/api";

import { createContext } from "~/trpc/server";
import { CreatePostForm, PostList } from "./components/posts";
import Spinner from "./components/spinner";

async function Posts() {
  const helpers = createServerSideHelpers({
    router: appRouter,
    ctx: await createContext(),
    transformer: superjson,
  });
  await helpers.post.all.prefetch();
  const dehydratedState = dehydrate(helpers.queryClient);

  return (
    <HydrationBoundary state={dehydratedState}>
      <PostList />
    </HydrationBoundary>
  );
}

export default async function HomePage() {
  return (
    <>
      <CreatePostForm />
      <Suspense fallback={<Spinner />}>
        <div className="w-full max-w-2xl overflow-y-scroll">
          <Posts />
        </div>
      </Suspense>
    </>
  );
}
```

**Reply by DavideCarvalho** [reply]
Oh I didn't know serverSideHelpers gave us access to queryClient instance, that's awesome! I'm going to try, ty!

**Reply by DavideCarvalho** [reply]
Just by importing "createServerSideHelpers" from "@trpc/react-query/server" I keep having the error "Attempted import error: 'createContext' is not exported from 'react' (imported as 'React')."

![image](https://github.com/trpc/trpc/assets/15630886/72869837-5466-42e2-a914-e75bc2acf888)

Could you send you dependencies versions?

**Reply by dihmeetree** [reply]
Which version of tRPC are you on? I believe it's only available in newer versions (10.x+ maybe?)

**Reply by DavideCarvalho** [reply]
From create t3 turbo repo - it's the repo I'm using as template - they are using trpc on version "11.0.0-next-beta.236"

![image](https://github.com/trpc/trpc/assets/15630886/9af65c08-c61c-4a2b-aca9-6537af9492fa)

**Reply by dihmeetree** [reply]
I'm on `11.0.0-next-beta.242` if that helps. You can import it like so:
```ts
import { createServerSideHelpers } from "@trpc/react-query/server";
```

**Reply by niklaswallerstedt** [reply]
I'm on `11.0.0-next-beta.234`, have you also bumped react query to latest?

**Reply by DavideCarvalho** [reply]
Just updated react query dependencies to the latest one available, but the error remains

![image](https://github.com/trpc/trpc/assets/15630886/4d0e5965-e87b-424b-9923-8ce42828642f)

![image](https://github.com/trpc/trpc/assets/15630886/bddc92cd-a1ce-4ff4-b6a5-c6cf11c8f62f)

**Comment by ryanagillie** [comment]
Hey all! Found an even better method for v10 (and tq@v4), just takes a tiny custom wrapper that wrapps the official TQ `<Hydrate>`. 

Found out the underlying problem was that [`createServerSideHelpers`'s `dehydrate()` automatically serializes the state for you](https://github.com/trpc/trpc/blob/main/packages/react-query/src/ssg/ssg.ts#L126), which breaks what tanstack is expecting (if you use superjson, which most people will). We just need to deserialize it and throw it back in.
```tsx
"use client";

import {
  Hydrate as TQHydrate,
  type HydrateProps as TQHydrateProps,
} from "@tanstack/react-query";

import tRPC from "./tRPC";

export type Props = TQHydrateProps & {
  state: Parameters<(typeof tRPC)["useDehydratedState"]>[1];
};
export default function Hydrate({ state, ...props }: Props) {
  return (
    <TQHydrate
      state={tRPC.useDehydratedState(tRPC.useUtils().client, state)}
      {...props}
    />
  );
}
```
and use it like
```tsx
  import { Hydrate } from 'path/to/wherever/you/exported/it';

  const helpers = createServerSideHelpers({
    router: appRouter,
    ctx: await createContext(),
    transformer: superjson,
  });
  await helpers.post.all.prefetch();

  return (
    <Hydrate state={helpers.dehydrate()}>
      <PostList />
    </HydrationBoundary>
  );
```

**Reply by DavideCarvalho** [reply]
I'll try with tq@5 and with beta versions of trpc tomorrow and I'll let you know my results!

But it sure looks like a good way of working with it. If it works, I'll surely use it while [the official concept of this is on the making](https://github.com/t3-oss/create-t3-turbo/pull/877)

**Reply by sebasptsch** [reply]
you can use https://trpc.io/docs/client/react/createTRPCQueryUtils in v11 to make the calls using the raw client.

**Reply by ryanagillie** [reply]
Not sure if it'll work on v11 and tq@5, looks like [they've removed the useHydrate hook](https://tanstack.com/query/latest/docs/framework/react/guides/migrating-to-v5#hydration-api-changes) (on my end the link isn't jumping, but if you ctrl+f to "Hydration API changes", you should find it)

**Reply by sebasptsch** [reply]
Docs about how to use with pages https://tanstack.com/query/latest/docs/framework/react/guides/advanced-ssr

**Reply by ryanagillie** [reply]
@KATT looks like the new changes to the transformer being on the link instead of the object itself makes wrapping the tq@5 `HydrationBoundary` with something like `<HydrationBoundary state={tRPC.useUtils().client.runtime.transformer.deserialize(state)} />` not possible as runtime is empty, any ideas?

**Reply by ryanagillie** [reply]
@DavideCarvalho updated my answer with a better implementation that wraps TQ's `<Hydrate>` instead of doing `useHydrate` ourselves

**Reply by ryanagillie** [reply]
Looks like v11 remove `useDehydratedState` as well, gonna have to keep looking

**Comment by dihmeetree** [comment]
> you can use https://trpc.io/docs/client/react/createTRPCQueryUtils in v11 to make the calls using the raw client.

Mind explaining this a bit? What does `createTRPCQueryUtils` achieve exactly, in the context of this PR? 🙂

**Reply by sebasptsch** [reply]
https://tanstack.com/query/latest/docs/framework/react/guides/ssr#full-nextjs-pages-router-example - and the other stuff on the page

**Reply by sebasptsch** [reply]
you can use the same approach but with the trpcQueryUtils to preload the query

https://tanstack.com/query/latest/docs/framework/react/guides/advanced-ssr

**Comment by Maxservais** [comment]
Anyone has got any update on what's the best way to deal with this?

Passing data from a Server component to a client component as initialData works fine but it's not 100% ideal and I would rather use the Hydration Boundaries if it's possible.

Cheers!

**Reply by DavideCarvalho** [reply]
I've found this question yesterday: https://www.answeroverflow.com/m/1169761966248693810. Haven't tested tho

**Reply by DavideCarvalho** [reply]
Since RSC, tRPC layer is thinner than ever, so maybe wrapping our components with tanstack hydration layer directly is the easiest solution

**Reply by DavideCarvalho** [reply]
I got it! The example from this post didn't work for me, but with some small tweaks I got it working! Yeah, unfortunately it requires some boilerplate code, but IMO it's better than passing down initialProps

<img width="1508" alt="image" src="https://github.com/trpc/trpc/assets/15630886/9d16c0a2-3ace-4161-8eac-f1c3b1455864">

**Reply by Maxservais** [reply]
Thank you, it works fine!
